examples/cpp/raw_mandelbrot.cpp

```diff
@@ -0,0 +1,175 @@
+        pos_params[ZCENTER], pos_params[WCENTER]
+    );
+    dmat4 rot_matrix = rotated_matrix(pos_params);
+    rot_matrix = rot_matrix / im->totalXres();
+    dvec4 deltax = rot_matrix[VX];
+    dvec4 deltay = rot_matrix[VY];
+    dvec4 delta_aa_x = deltax / 2.0;
+    dvec4 delta_aa_y = deltay / 2.0;
+    dvec4 topleft = center - deltax * im->totalXres() / 2.0 - deltay * im->totalYres() / 2.0;
+    topleft += delta_aa_x + delta_aa_y;
+
+    int w = im->Xres();
+    int h = im->Yres();
+    // we put these variables out of the loop scope to use its previous value
+    int iters_taken = 0;
+    int min_period_iters = 0;
+    // now we calculate every pixel (for a 2D image projection in a single tile)
+    for (auto x = 0; x < w; x++) {
+        for (auto y = 0; y < h; y++) {
+            // calculate the position in 4D (x, y, z, w)
+            dvec4 pos = topleft + x * deltax + y * deltay;
+            // run the formula
+            double dist = 0.0;
+            int fate = 0;
+            int solid = 0;
+            int direct_color = 0;
+            double colors[4] = {0.0};
+            int inside = 0;
+            if (iters_taken == -1) { // we got inside the last time so we'll probably do it again
+                min_period_iters = 0;
+            } else {
+                min_period_iters += 10;
+            }
+            pf_handle->vtbl->calc(
+                pf_handle,
+                pos.n,
+                100, // max iters
+                -1, // wrap param
+                min_period_iters, // min period iters (half the max)
+                1.0E-9, // period tolerance
+                x, y, 0, // x, y and aa: these values are not needed in the formula but required as arguments for debugging purposes
+                &iters_taken, &fate, &dist, &solid,
+                &direct_color, &colors[0]);
+            // process the formula output and get the value from colormap
+            rgba_t color;
+            if (fate & FATE_INSIDE) {
+                iters_taken = -1;
+                inside = 1;
+            }
+            if (direct_color) {
+                color = cmap->lookup_with_dca(solid, inside, colors);
+                fate |= FATE_DIRECT;
+            } else {
+                color = cmap->lookup_with_transfer(dist, solid, inside);
+            }
+            if (solid)
+            {
+                fate |= FATE_SOLID;
+            }
+            // this is only needed for optimization and zooming
+            im->setIter(x, y, iters_taken);
+            im->setFate(x, y, 0, fate);
+            im->setIndex(x, y, 0, dist);
+            // put the pixel color into the image buffer position
+            im->put(x, y, color);
+        }
+    }
+
+    // save the image
+    FILE *image_file = fopen("./output/mandelbrot.png", "wb");
+    image_file_t image_file_type = FILE_TYPE_PNG;
+    ImageWriter *image_writer = ImageWriter::create(image_file_type, image_file, im);
```

Comment:
`std::unique_ptr<ImageWriter> image_writer{ImageWriter::create(...)};` would be more ideal with the automatic memory management this gives you.